File: Leetcode/Python3/78. Subsets/solution.py
# https://leetcode.com/problems/subsets/


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        answer = []

        def dfs(result, index):
            answer.append(result[:])
            for idx in range(index, len(nums)):
                result.append(nums[idx])
                dfs(result, idx + 1)
                result.pop()

        dfs([], 0)
        return answer

File: Leetcode/Typescript/78. Subsets/solution.ts
// https://leetcode.com/problems/subsets/description/

function subsets(nums: number[]): number[][] {
  const result: number[][] = [];
  function dfs(prev: number[], index: number) {
    result.push(prev.slice());
    if (prev.length === nums.length) {
      return;
    }

    for (let idx = index; idx < nums.length; idx++) {
      prev.push(nums[idx]);
      dfs(prev, idx + 1);
      prev.pop();
    }
  }

  dfs([], 0);
  return result;
}
